### PR TITLE
[None][fix] Fix encoder-decoder beam search corruption via per-slot fragmentPointerDevice

### PR DESCRIPTION
**Pull request by achartier**
@coderabbitai summary

## Description

Fixes two bugs in the TRT backend (`ModelRunnerCpp`) that caused corrupted output when concurrent requests with different beam widths (e.g. beam=1 and beam=5) were submitted to the same executor on encoder-decoder models (Whisper) with beam search.

**Bug 1 — `mInflightReqIds` leak on mixed-beam exception**

When concurrent requests with different beam widths triggered the `verifyRequests()` exception in `TrtGptModelInflightBatching::forwardAsync`, the exception handler freed the KV-cache but did not remove request IDs from `mInflightReqIds`. The subsequent `changeBeamWidth()` call asserts this set is empty and aborted.

Fix: erase affected IDs from `mInflightReqIds` and call `changeBeamWidth(mOperatingBeamWidth)` in the exception catch block.

**Bug 2 — `gather_generation_logits` corruption via shared `fragmentPointerDevice`**

`mergeLogitsFragmentsKernel` reads fragment GPU addresses from `fragmentPointerDevice` (shape `[kCACHE_LENGTH]`). When multiple requests in the same batch flush their logits fragments sequentially, each overwrites that shared device buffer before the previous kernel has committed — a race causing intermittent degenerate output with `gather_generation_logits=True`.

Fix: resize `fragmentPointerDevice` from `[kCACHE_LENGTH]` to `[maxBatchSize, kCACHE_LENGTH]`, matching the layout of `fragmentPointerHost`. Add `getFragmentPointerSlot()` which returns both host and device rows atomically and advances `workIdx` once, making call-order errors structurally impossible. Each request in the batch now has its own device-side pointer array.

## Test Coverage

**C++ unit test** (`cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp`):
- `CopyGenerationLogitsTest/KernelMergePathProducesCorrectHostLayoutAndSlotsAreIsolated` — verifies Bug 2 fix: each batch slot gets its own device-side pointer row, and slots are fully isolated from one another.

**Validation on Whisper large-v3 with concurrent mixed-beam execution:**
| Configuration | Before fixes | After fixes |
|---|---|---|
| beam=5 only, no mixing | — | PASS 10/10 |
| mix + no `gather_generation_logits` | crash/corrupt | **PASS 20/20** |
| mix + `gather_generation_logits=True` | crash/corrupt | **PASS 50/50** |

## PR Checklist

- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment \`/bot help\`.

### CODE REVIEW

File: cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+// places the same physical block IDs in every beam slot.
+//
+// This tests the isCrossKv() branch added by fix 1: when all beams share the
+// same encoder output, the output offset table must reflect that.  In the
+// production Whisper case the allocator gives each beam its own physical
+// blocks, so beams 1..N-1 would reference uninitialised GPU memory without
+// the fix.  This unit test uses a simple context-only sequence where the
+// allocator happens to share blocks across beams; the observable property
+// (all beam slots equal beam-0) still holds and constitutes a sanity check.
+TEST(CrossKvBeamSharingTest, CopyBlockOffsetsAllBeamsShareBeam0Blocks)
```
[the remark this answers is not in the document]
Comment (achartier) [reply cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp]: fixed by 92ff8bd5dbeed319bd10ecce1098b67150cd2208

File: cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        }
+        llmReq->addGenerationLogitsFragment(frag);
+    }
+    ASSERT_EQ(llmReq->getGenerationLogitsFragmentsSize(), numSteps);
+
+    // Dummy cache — not accessed by the direct-copy implementation.
+    RuntimeBuffers::GenerationLogitsCache dummyCache;
+
+    utils::copyGenerationLogits(dummyCache, bufferMgr, *llmReq, /*beforeDecoder=*/false, /*numDroppedTokens=*/{});
+
```
[the remark this answers is not in the document]
Comment (achartier) [reply cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp]: fixed by 92ff8bd5dbeed319bd10ecce1098b67150cd2208

File: cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp [hunk 3]
```diff
@@ -3957,16 +3957,21 @@ SizeType32 KVCacheManager::copyBlockOffsets(ITensor& output, SizeType32 outputSl
         {
             for (SizeType32 beamIdx = 0; beamIdx < beamWidth; ++beamIdx)
             {
-                auto const beamBlockCount = cacheBlockIds[beamIdx].size();
-                auto const copyChunkSize = beamBlockCount * sizeof(tk::KVCacheIndex);
+                // For cross-KV (encoder features), all beams of a request share the same encoder output.
+                // Always use beam 0's block IDs so all beams attend to the correct encoder features.
```
Comment (Funatiq) [inline cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp]: Don't we have the same situation for normal KV cache? I think there we fill all beams with the same block IDs for the context length. Why don't we do the same for cross-KV? Or do the `cacheBlockIds[beamIdx]` for `beamIdx > 1` simply not exist for cross-KV?

File: cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp [hunk 4]
```diff
@@ -3957,16 +3957,21 @@ SizeType32 KVCacheManager::copyBlockOffsets(ITensor& output, SizeType32 outputSl
...
+                auto const srcBeamIdx = isCrossKv() ? 0 : beamIdx;
+                TLLM_CHECK_WITH_INFO(!isCrossKv() || !cacheBlockIds.empty(),
+                    "Cross-KV sequence has no block IDs for request %lu", requestId);
+                auto const effectiveBlockCount = isCrossKv() ? cacheBlockIds[0].size() : cacheBlockIds[beamIdx].size();
```
Comment (Funatiq) [inline cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp]:
```suggestion
                auto const effectiveBlockCount = cacheBlockIds[srcBeamIdx].size();
```

File: cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Regression tests for encoder-decoder beam-search fixes.
+//
+// Fix 1 — cross-KV copyBlockOffsets shares beam-0 blocks across all beams.
```
Comment (Funatiq) [inline cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp]: Saying "Fix 1" and "Fix 3" here does only make sense in the context of the PR. It doesn't add any value to the test file. I would also suggest removing the summary from the top of the file. I don't think it's something that we would want to maintain. You could move it to the individual tests if you think it is useful.

File: cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp (hunk 3 above)
Comment (achartier) [reply cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp]: Thanks for the catch. It looks like Claude confused itself by printing the IDs as 64b while they are 32b and incorrectly concluded there was corruption happening. What was referred to as 'fix 1' is not needed, so I reverted the code change.

File: cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp (hunk 4 above)
Comment (achartier) [reply cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp]: That was part of the code reverted.

File: cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp (hunk 5 above)
Comment (achartier) [reply cpp/tests/unit_tests/batch_manager/encDecBeamSearchTest.cpp]: Cleaned up in https://github.com/NVIDIA/TensorRT-LLM/pull/15461/commits/29277dd1e63d1ebb7b579bc6c562f9c0094bd2a0

File: cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp [hunk 6]
```diff
@@ -3957,16 +3957,16 @@ SizeType32 KVCacheManager::copyBlockOffsets(ITensor& output, SizeType32 outputSl
         {
             for (SizeType32 beamIdx = 0; beamIdx < beamWidth; ++beamIdx)
             {
-                auto const beamBlockCount = cacheBlockIds[beamIdx].size();
-                auto const copyChunkSize = beamBlockCount * sizeof(tk::KVCacheIndex);
+                auto const effectiveBlockCount = cacheBlockIds[beamIdx].size();
```
Comment (Funatiq) [inline cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp]: This could also be reverted.

**Review by Funatiq — APPROVED** [review]
